### PR TITLE
Check ELF file for errors before deploy

### PR DESCRIPTION
**Pull request by alexfmpe**
#### Problem
Checking the ELF file is valid requires actually uploading to a node - this has two disadvantages:
- a running node is required
- the error message from the ELF check is mixed with the node logs, and the CLI receives an unspecific error message 
```
Error: dynamic program error: Finalizing program account failed: Error processing Instruction 0: invalid account data for instruction
```

#### Summary of Changes
Check the ELF file for errors before uploading and exit immediately if so.

```
$ cargo run --bin solana deploy program.so 
   Compiling solana-cli v1.5.0 
    Finished dev [unoptimized + debuginfo] target(s) in 8.73s
     Running `target/debug/solana deploy program.so`
Error: dynamic program error: ELF error: ELF error: Failed to parse ELF file: read-write: bad offset 441936
```

This was [originally](https://github.com/obsidiansystems/solana/commit/db2f8ec4fc7b9ccbfdc68ace67d767dbac9330dd) done on top of v1.3.9 - would a PR to v.1.3.9 be useful?

### CODE REVIEW

File: cli/src/cli.rs [hunk 1]
```diff
@@ -1144,6 +1146,13 @@ fn process_deploy(
         CliError::DynamicProgramError(format!("Unable to read program file: {}", err))
     })?;
 
+    let exe = EbpfVm::create_executable_from_elf(
+        &program_data,
+        Some(|x| bpf_verifier::check(x, true)));
+    exe.map_err(|err| {
+        CliError::DynamicProgramError(format!("ELF error: {}", err))
+    })?;
+
```
Comment (mvines) [inline cli/src/cli.rs:1155]: I think we could just do this:
```suggestion
    let _ = EbpfVm::create_executable_from_elf(
        &program_data,
        Some(|x| bpf_verifier::check(x, true))).map_err(|err| {
        CliError::DynamicProgramError(format!("ELF error: {}", err))
    })?;

```

**Review by mvines — COMMENTED** [review]
Nice idea!   

@jackcmay - wdyt?

**Comment by jackcmay** [comment]
This is a great idea

**Comment by mvines** [comment]
> This was originally done on top of v1.3.9 - would a PR to v.1.3.9 be useful?

PR into `master` is perfect.   We'll backport to v1.3 using mergify once it lands

**Comment by mvines** [comment]
Mind running `cargo fmt` over your PR?  CI is complaining about that

File: cli/Cargo.toml [hunk 2]
```diff
@@ -36,6 +37,7 @@ solana-config-program = { path = "../programs/config", version = "1.5.0" }
 solana-faucet = { path = "../faucet", version = "1.5.0" }
 solana-logger = { path = "../logger", version = "1.5.0" }
 solana-net-utils = { path = "../net-utils", version = "1.5.0" }
+solana_rbpf = "=0.1.31"
```
Comment (jackcmay) [inline cli/Cargo.toml:40]: fyi, pending PR will require this be bumped to v01.1.32:

https://github.com/solana-labs/solana/pull/12728

**Comment by jackcmay** [comment]
@alexfmpe Looks like CI is failing due to formatting.  Try running `cargo fmt --all` and resubmitting

**Comment by mvines** [comment]
@jackcmay - this is ready to land, how do you want to manage the merge conflict with https://github.com/solana-labs/solana/pull/12728?

**Comment by jackcmay** [comment]
Will do!

**Comment by alexfmpe** [comment]
I can probably push a bump tomorrow.

**Pull request merged**